src/EntityFramework.Storage/Stores/ClientStore.cs

```diff
@@ -50,23 +51,32 @@ public ClientStore(IConfigurationDbContext context, ILogger<ClientStore> logger)
         /// </returns>
         public virtual async Task<Duende.IdentityServer.Models.Client> FindClientByIdAsync(string clientId)
         {
-            var query = Context.Clients
-                        .Where(x => x.ClientId == clientId)
-                        .Include(x => x.AllowedCorsOrigins)
-                        .Include(x => x.AllowedGrantTypes)
-                        .Include(x => x.AllowedScopes)
-                        .Include(x => x.Claims)
-                        .Include(x => x.ClientSecrets)
-                        .Include(x => x.IdentityProviderRestrictions)
-                        .Include(x => x.PostLogoutRedirectUris)
-                        .Include(x => x.Properties)
-                        .Include(x => x.RedirectUris)
-                        .AsNoTracking(); 
-            
-            var client = await query.SingleOrDefaultAsync();
-            if (client == null) return null;
+            var results = await Context.Clients
+                .Where(x => x.ClientId == clientId)
+                .Include(x => x.AllowedCorsOrigins)
+                .Include(x => x.AllowedGrantTypes)
+                .Include(x => x.AllowedScopes)
+                .Include(x => x.Claims)
+                .Include(x => x.ClientSecrets)
+                .Include(x => x.IdentityProviderRestrictions)
+                .Include(x => x.PostLogoutRedirectUris)
+                .Include(x => x.Properties)
+                .Include(x => x.RedirectUris)
+                .AsNoTracking()
+                .ToArrayAsync();
 
-            var model = client.ToModel();
+            Client entity;
+            try
+            {
+                entity = results.SingleOrDefault(x => x.ClientId == clientId);
+            }
+            catch (InvalidOperationException)
+            {
+                Logger.LogError("Duplicate Client entries found for clientId {clientId}", clientId);
+                return null;
+            }
```

Comment:
How about checking the `Length` property instead of checking with exception?
```suggestion
            if (results.Length == 0)
            {
                return null;            
            }
            
            if (results.Length > 1)
            {
                Logger.LogError("Duplicate Client entries found for clientId {clientId}", clientId);
                return null;            
            }
            
            var model = results[0].ToModel();
```